Add default subcommand to commands

Eveldee opened the pull request:
This allows one and only one subcommand to be marked as **default** in a command.
When a subcommand is marked as **default** (see DefaultExampleCommands.cs), it has the following effects:
- If no subcommand is specified it will fallback to the **default** one without any args
- If an unknown subcommand is specified, it will fallback to the **default** one with everything after the parent command as args

Tell me if you want to change how the **default** subcommand is set, I first tried with a "DefaultSubCommand" Attribute, but it felt wrong so I did it with another constructor for the SubCommand Attribute.

I also fixed a little issue with tests as my CultureInfo use ',' instead of '.' for the decimal separator, now it works with both, using the CurrentCulture one.

Eveldee committed:
Fix test on systems with different culture

Eveldee committed:
Add IsDefault to SubCommand

If there's a defaut SubCommand, it will fallback to it if no or an invalid command is invoked

Eveldee committed:
Add tests for default SubCommand

ZakFahey commented on EasyCommands/EasyCommands/Commands/CommandGroupDelegate.cs:
```diff
@@ -34,6 +35,16 @@ public CommandGroupDelegate(Context<TSender> context, string mainName, string[]
+                    {
+                        if (defaultCommand != null)
+                        {
+                            throw new CommandRegistrationException($"There is two or more default sub commands in {command.Name}.");
```
Grammar: should be "there are two"

ZakFahey commented on EasyCommands/EasyCommands/Attributes.cs:
```diff
@@ -24,8 +24,19 @@ public Command(params string[] names)
+        public SubCommand(bool isDefault, params string[] names) : base(names)
+        {
+            IsDefault = isDefault;
+        }
```
Can you add an override with zero arguments that acts as the default subcommand and has no names? You may also want to double check that it doesn't clash with a zero-length params array.

ZakFahey commented on EasyCommands/EasyCommands/Commands/CommandGroupDelegate.cs:
```diff
@@ -51,18 +62,42 @@ public override void Invoke(TSender sender, string args)
-            subcommands[subcommand].Invoke(sender, subcommandArgs);
         }
 
         public string SubcommandList()
```
Can you make sure that any user messages (error messages, the HelpCommand in the Example project) generated by this system show reasonable looking messages for default subcommands? I believe you'd have to change syntaxDocumentation in BaseCommandDelegate.

ZakFahey commented:
I'm also not sure if there's any point to there being a named failover for default commands. It should just be what happens when you have no name argument in the attribute. Nobody's going to type a more verbose alt alias of a command.

ZakFahey commented:
Also, could you document this functionality in the documentation?

Eveldee commented:
I tried naming a command with an empty stirng: "" and it worked well with documentation using EasyCommandsTShock on a Terraria server so I guess if I do that, it will be good.
I'm going to try doing this

Eveldee commented:
I just realized when doing the tests that the empty constructor is used for inferring the command name from the method name, should I create a constructor with only one bool argument instead for the default one? So if it's true, it will se the name to "" and if it's false, it will just do the same as no argument.

Something like this:
```cs
public SubCommand(bool isDefault) : base(isDefault ? new string[] { "" } : Array.Empty<string>())
{
	IsDefault = isDefault;
}
```

ZakFahey commented:
I think you can go without the boolean argument altogether and just say that if there is no name, it's a default command. If method overriding doesn't work because C# interprets it as a zero-length params array, you can just check if length is zero there instead of having another constructor.

ZakFahey commented:
An attribute with one isDefault should be okay, but it might encourage people to do `[SubCommand(true)]` instead of `[SubCommand(isDefault: true)]`. Both compile but one is less readable. I'd maybe consider a format that forces you to be explicit in how it shows, such as with an enum argument or another `DefaultSubCommand` attribute class.

With a subcommand of "" I think that it might look okay at first glance but it would have an extra space. You'd see `/command<two spaces>arg arg` (GitHub formats it away so I can't show it...)

Eveldee commented:
I have to give a name to the command so "" seems good for me, and I need a way to recognize between an empty constructor which should infer the command name from method name or for a default subcommand.
Using an enum is a good idea if you want it to make it explicit, I'm going to try it

Eveldee commented:
I have this so far:
```cs
/// <summary>
/// Change how a <see cref="SubCommand"/> is executed
/// </summary>
public enum SubCommandType
{
	/// <summary>
	/// This does nothing special, it just execute the command when it's called
	/// </summary>
	Standard,
	/// <summary>
	/// Mark the command as the default one, it will be executed whenever an invalid command is executed
	/// </summary>
	Default
}
```
```cs
/// <param name="subCommandType">Determine whether this subcommand is a default subcommand or not</param>
public SubCommand(SubCommandType subCommandType) : base(subCommandType == SubCommandType.Default ? new string[] { "" } : Array.Empty<string>())
{
	IsDefault = subCommandType == SubCommandType.Default;
}
```

As for the name, I think I should let "" so the argument parser works well, I just need to update the syntaxDocumentation to not print two spaces

Eveldee commented:
How should the help display the command ?
Currently, it's:
```
default <hello|bye|>
```

It may be better like this:
```
default [<hello|bye>]
```
Or like this:
```
default [hello|bye]
```

ZakFahey commented:
Something like `default <hello|bye>`. Omit the default from the subcommand list.

Eveldee committed:
Change how a default SubCommand is created

Now use a constructor with an enum

Eveldee committed:
Fix how SyntaxDocumention handle default commands

There is no longer two whitespaces
The default command is no longer shown in <sub1|sub2|default> syntax documentation

Eveldee committed:
Update Example and Tests

Eveldee committed:
Add Documentation for default subcommands

Eveldee committed:
Update and Change root of Documentation

Eveldee committed:
Add language to codeblocks

Eveldee commented:
I've made all the changes you requested, I tried it in real conditions and it worked well.
I also made some little changes to the documentation to make it more readable.

I now need to investigate on the other repository why the `[AllowServer]` attribute does not work with TShock, I'll open a PR as I did for this one if I find the answer.

Eveldee commented:
Here is a little example using it on my Terraria server with the following command:
```cs
[Command("vote", "v")]
[HelpText("Vote for different things.")]
public class VoteCommand : CommandCallbacks<TSPlayer>
{
	[SubCommand("time", "t")]
	[HelpText("Vote for a time set.")]
	public void Time([AllowSpaces] string time)
	{
		// Some code
	}

	[SubCommand("rain", "r")]
	[HelpText("Vote for a rain on/off toggle")]
	public void Rain()
	{
		// Some code
	}

	[SubCommand(SubCommandType.Default)]
	[HelpText("Vote for the last request.")]
	[AllowServer(false)]
	public void Vote()
	{
		// Some code
	}
}
```

Result:
```
help vote
Server executed: /help vote.
/vote help:
Vote for different things. Syntax: /vote <time|rain>
: vote abc
Server executed: /vote abc.
Incorrect number of arguments! Proper syntax: /vote
: vote time
Server executed: /vote time.
Incorrect number of arguments! Proper syntax: /vote time <time>
: vote time 12:45
Server executed: /vote time 12:45.
A vote has started for: "Setting time to 12:45", type "/vote" to vote
Vote status: 0/1
: vote
Server executed: /vote.
Server voted, vote status: 1/1
Vote has succeeded, executing request...
Time set to 12:45
```

Also note that the server can still vote even if there is `[AllowServer(false)]`, btu that's an issue for the other repository.

ZakFahey commented:
Ok, I'll need to make a Nuget release and then have a PR to reference the newer version in the TShock library.
